File: dart-sdks/lib/src/sui/transactions.dart
import 'dart:convert';
import 'dart:typed_data';

const String systemStateObjectId = '0x5';
const String stakeRequestTarget = '0x3::sui_system::request_add_stake';
const String unstakeRequestTarget = '0x3::sui_system::request_withdraw_stake';

class ObjectRef {
  const ObjectRef({required this.objectId, required this.digest, required this.version});

  final String objectId;
  final String digest;
  final int version;
}

class Inputs {
  static Map<String, dynamic> pure(List<int> value) {
    return {
      '\$kind': 'Pure',
      'Pure': {'bytes': base64Encode(value)},
    };
  }

  static Map<String, dynamic> objectRef(ObjectRef ref) {
    return {
      '\$kind': 'Object',
      'Object': {
        '\$kind': 'ImmOrOwnedObject',
        'ImmOrOwnedObject': {
          'objectId': ref.objectId,
          'digest': ref.digest,
          'version': ref.version,
        },
      },
    };
  }

  static Map<String, dynamic> sharedObjectRef(String objectId, bool mutable, int initialSharedVersion) {
    return {
      '\$kind': 'Object',
      'Object': {
        '\$kind': 'SharedObject',
        'SharedObject': {
          'objectId': objectId,
          'mutable': mutable,
          'initialSharedVersion': initialSharedVersion,
        },
      },
    };
  }
}

class TransactionCommands {
  static Map<String, dynamic> moveCall(
    String target,
    List<Map<String, dynamic>> args, [
    List<String>? typeArgs,
  ]) {
    final segments = target.split('::');
    final package = segments.isNotEmpty ? segments[0] : '';
    final module = segments.length > 1 ? segments[1] : '';
    final function = segments.length > 2 ? segments[2] : '';

    return {
      '\$kind': 'MoveCall',
      'MoveCall': {
        'package': package,
        'module': module,
        'function': function,
        'arguments': args,
        'typeArguments': typeArgs ?? <String>[],
      },
    };
  }

  static Map<String, dynamic> splitCoins(Map<String, dynamic> coin, List<Map<String, dynamic>> amounts) {
    return {
      '\$kind': 'SplitCoins',
      'SplitCoins': {'coin': coin, 'amounts': amounts},
    };
  }

  static Map<String, dynamic> transferObjects(List<Map<String, dynamic>> objects, Map<String, dynamic> address) {
    return {
      '\$kind': 'TransferObjects',
      'TransferObjects': {'objects': objects, 'address': address},
    };
  }

  static Map<String, dynamic> mergeCoins(Map<String, dynamic> destination, List<Map<String, dynamic>> sources) {
    return {
      '\$kind': 'MergeCoins',
      'MergeCoins': {'destination': destination, 'sources': sources},
    };
  }

  static Map<String, dynamic> publish(List<List<int>> modules, List<String> dependencies) {
    return {
      '\$kind': 'Publish',
      'Publish': {
        'modules': modules.map(base64Encode).toList(),
        'dependencies': dependencies,
      },
    };
  }

  static Map<String, dynamic> upgrade(
    List<List<int>> modules,
    List<String> dependencies,
    String packageId,
    Map<String, dynamic> ticket,
  ) {
    return {
      '\$kind': 'Upgrade',
      'Upgrade': {
        'modules': modules.map(base64Encode).toList(),
        'dependencies': dependencies,
        'package': packageId,
        'ticket': ticket,
      },
    };
  }

  static Map<String, dynamic> makeMoveVec(String? typeTag, List<Map<String, dynamic>> elements) {
    return {
      '\$kind': 'MakeMoveVec',
      'MakeMoveVec': {'type': typeTag, 'elements': elements},
    };
  }
}

class TransactionData {
  String sender = '';
  dynamic expiration;
  final Map<String, dynamic> gasData = <String, dynamic>{};
  final List<Map<String, dynamic>> inputs = <Map<String, dynamic>>[];
  final List<Map<String, dynamic>> commands = <Map<String, dynamic>>[];
}

class Transaction {
  Transaction({this.client});

  final TransactionData data = TransactionData();
  final dynamic client;

  List<Map<String, dynamic>> get commands => data.commands;
  List<Map<String, dynamic>> get inputs => data.inputs;

  void setSender(String sender) {
    data.sender = sender;
  }

  void setSenderIfNotSet(String sender) {
    if (data.sender.isEmpty) {
      data.sender = sender;
    }
  }

  void setGasBudget(int budget) {
    data.gasData['budget'] = budget.toString();
  }

  void setGasBudgetIfNotSet(int budget) {
    if (data.gasData['budget'] == null) {
      setGasBudget(budget);
    }
  }

  void setExpiration(dynamic expiration) {
    data.expiration = expiration;
  }

  void setGasPrice(int price) {
    data.gasData['price'] = price.toString();
  }

  void setGasOwner(String owner) {
    data.gasData['owner'] = owner;
  }

  void setGasPayment(List<Map<String, dynamic>> payments) {
    data.gasData['payment'] = payments;
  }

  Map<String, dynamic> gas() => {
        '\$kind': 'GasCoin',
        'GasCoin': true,
      };

  Map<String, dynamic> addInput(Map<String, dynamic> arg) {
    data.inputs.add(arg);
    return {
      '\$kind': 'Input',
      'Input': data.inputs.length - 1,
    };
  }

  Map<String, dynamic> object(dynamic value) {
    if (value is String) {
      return addInput({
        '\$kind': 'UnresolvedObject',
        'UnresolvedObject': {'objectId': value},
      });
    }
    if (value is Map<String, dynamic> && value['\$kind'] == 'Input') {
      return value;
    }
    return addInput(Map<String, dynamic>.from(value as Map));
  }

  Map<String, dynamic> pure(List<int> value) {
    return addInput(Inputs.pure(value));
  }

  Map<String, dynamic> addCommand(Map<String, dynamic> cmd) {
    data.commands.add(cmd);
    return {
      '\$kind': 'Result',
      'Result': data.commands.length - 1,
    };
  }

  Map<String, dynamic> moveCall(String target, List<Map<String, dynamic>> args, [List<String>? typeArgs]) {
    return addCommand(TransactionCommands.moveCall(target, args, typeArgs));
  }

  Map<String, dynamic> transferObjects(List<Map<String, dynamic>> objects, Map<String, dynamic> address) {
    return addCommand(TransactionCommands.transferObjects(objects, address));
  }

  Map<String, dynamic> splitCoins(Map<String, dynamic> coin, List<Map<String, dynamic>> amounts) {
    return addCommand(TransactionCommands.splitCoins(coin, amounts));
  }

  Map<String, dynamic> mergeCoins(Map<String, dynamic> destination, List<Map<String, dynamic>> sources) {
    return addCommand(TransactionCommands.mergeCoins(destination, sources));
  }

  Map<String, dynamic> publish(List<List<int>> modules, List<String> dependencies) {
    return addCommand(TransactionCommands.publish(modules, dependencies));
  }

  Map<String, dynamic> upgrade(
    List<List<int>> modules,
    List<String> dependencies,
    String packageId,
    Map<String, dynamic> ticket,
  ) {
    return addCommand(TransactionCommands.upgrade(modules, dependencies, packageId, ticket));
  }

  Map<String, dynamic> publishUpgrade(
    List<List<int>> modules,
    List<String> dependencies,
    String packageId,
    Map<String, dynamic> ticket,
  ) {
    return upgrade(modules, dependencies, packageId, ticket);
  }

  Map<String, dynamic> customUpgrade(
    List<List<int>> modules,
    List<String> dependencies,
    String packageId,
    Map<String, dynamic> ticket,
  ) {
    return upgrade(modules, dependencies, packageId, ticket);
  }

  Map<String, dynamic> makeMoveVec(String? typeTag, List<Map<String, dynamic>> elements) {
    return addCommand(TransactionCommands.makeMoveVec(typeTag, elements));
  }

  Map<String, dynamic> transferSui(String recipient, int amount) {
    final splitResult = splitCoins(gas(), [pure(_u64Bytes(amount))]);
    return transferObjects([splitResult], pure(utf8.encode(recipient)));
  }

  Map<String, dynamic> splitCoinEqual(Map<String, dynamic> coin, {required int splitCount, required int amountPerSplit}) {
    if (splitCount <= 0) {
      throw ArgumentError.value(splitCount, 'splitCount', 'must be positive');
    }
    final amounts = List.generate(splitCount, (_) => pure(_u64Bytes(amountPerSplit)));
    return splitCoins(coin, amounts);
  }

  Map<String, dynamic> splitCoinAndReturn(Map<String, dynamic> coin, {required int amount, required String recipient}) {
    final splitResult = splitCoins(coin, [pure(_u64Bytes(amount))]);
    transferObjects([splitResult], pure(utf8.encode(recipient)));
    return splitResult;
  }

  Map<String, dynamic> stakeCoin({
    required List<dynamic> coins,
    required String validatorAddress,
    int? amount,
    String systemStateObject = systemStateObjectId,
  }) {
    final coinArgs = coins.map((coin) => object(coin)).toList();
    final coinsVec = makeMoveVec(null, coinArgs);
    final amountArg = pure(_optionU64Bytes(amount));

    return moveCall(
      stakeRequestTarget,
      [
        object(systemStateObject),
        coinsVec,
        amountArg,
        pure(utf8.encode(validatorAddress)),
      ],
    );
  }

  Map<String, dynamic> unstakeCoin({required dynamic stakedCoin, String systemStateObject = systemStateObjectId}) {
    return moveCall(
      unstakeRequestTarget,
      [object(systemStateObject), object(stakedCoin)],
    );
  }

  Uint8List build() {
    final payload = <String, dynamic>{
      'Sender': data.sender,
      'Expiration': data.expiration,
      'GasData': data.gasData,
      'Inputs': data.inputs,
      'Commands': data.commands,
    };
    return Uint8List.fromList(utf8.encode(jsonEncode(payload)));
  }

  String buildBase64() => base64Encode(build());

  String serialize() => utf8.decode(build());

  Map<String, dynamic> getTransactionData() {
    return <String, dynamic>{
      'Sender': data.sender,
      'Expiration': data.expiration,
      'GasData': data.gasData,
      'Inputs': data.inputs,
      'Commands': data.commands,
    };
  }

  Map<String, dynamic> deferredExecution() {
    return {'sender': data.sender, 'tx_bytes': buildBase64()};
  }

  Future<Map<String, dynamic>> execute({
    dynamic client,
    List<String>? signatures,
    Map<String, dynamic>? options,
  }) async {
    final activeClient = _resolveClient(client);
    final txBytes = buildBase64();

    if (signatures == null && options == null) {
      return _rpcCall(activeClient, 'sui_executeTransactionBlock', [txBytes]);
    }
    return _rpcCall(
      activeClient,
      'sui_executeTransactionBlock',
      [txBytes, signatures ?? <String>[], options ?? <String, dynamic>{}],
    );
  }

  Future<Map<String, dynamic>> inspectAll({dynamic client, String? sender}) {
    final activeClient = _resolveClient(client);
    final activeSender = sender ?? data.sender;
    return _rpcCall(activeClient, 'sui_devInspectTransactionBlock', [activeSender, buildBase64()]);
  }

  Future<Map<String, int>> inspectForCost({dynamic client, String? sender}) async {
    final result = await inspectAll(client: client, sender: sender);
    final effects = (result['result'] as Map<String, dynamic>?)?['effects'] as Map<String, dynamic>?;
    final gasUsed = effects?['gasUsed'] as Map<String, dynamic>?;

    final computation = int.parse('${gasUsed?['computationCost'] ?? 0}');
    final storage = int.parse('${gasUsed?['storageCost'] ?? 0}');
    final rebate = int.parse('${gasUsed?['storageRebate'] ?? 0}');

    return {
      'computation_cost': computation,
      'storage_cost': storage,
      'storage_rebate': rebate,
      'total_cost': computation + storage - rebate,
    };
  }

  static Transaction fromSerialized(String serialized) {
    final raw = serialized.startsWith('{') ? utf8.encode(serialized) : base64Decode(serialized);
    final payload = jsonDecode(utf8.decode(raw)) as Map<String, dynamic>;

    final tx = Transaction();
    tx.data.sender = '${payload['Sender'] ?? ''}';
    tx.data.expiration = payload['Expiration'];
    tx.data.gasData.addAll((payload['GasData'] as Map?)?.cast<String, dynamic>() ?? <String, dynamic>{});
    tx.data.inputs.addAll(((payload['Inputs'] as List?) ?? const <dynamic>[])
        .map((e) => Map<String, dynamic>.from(e as Map))
        .toList());
    tx.data.commands.addAll(((payload['Commands'] as List?) ?? const <dynamic>[])
        .map((e) => Map<String, dynamic>.from(e as Map))
        .toList());
    return tx;
  }

  dynamic _resolveClient(dynamic injected) {
    final active = injected ?? client;
    if (active == null) {
      throw StateError('client is required for execution and inspection');
    }
    return active;
  }

  Future<Map<String, dynamic>> _rpcCall(dynamic c, String method, List<dynamic> params) async {
    try {
      final out = await c.call(method, params);
      return Map<String, dynamic>.from(out as Map);
    } catch (_) {
      final out = await c.execute(method, params);
      return Map<String, dynamic>.from(out as Map);
    }
  }

  static List<int> _u64Bytes(int value) {
    if (value < 0) {
      throw ArgumentError.value(value, 'value', 'u64 value must be non-negative');
    }
    final max = (BigInt.one << 64) - BigInt.one;
    if (BigInt.from(value) > max) {
      throw RangeError.value(value, 'value', 'u64 value out of range');
    }
    final out = ByteData(8)..setUint64(0, value, Endian.little);
    return out.buffer.asUint8List();
  }

  static List<int> _optionU64Bytes(int? value) {
    if (value == null) {
      return const [0];
    }
    return <int>[1, ..._u64Bytes(value)];
  }
}

class ResolveContext {
  ResolveContext({required this.transaction});

  final Transaction transaction;
  final List<Map<String, dynamic>> unresolvedInputs = <Map<String, dynamic>>[];
}

typedef ResolvePlugin = Future<void> Function(ResolveContext context);

class ResolverPluginError implements Exception {
  ResolverPluginError({required this.index, required this.pluginName, required this.cause});

  final int index;
  final String pluginName;
  final Object cause;

  Map<String, dynamic> toJson() {
    return {
      'error_type': 'ResolverPluginError',
      'index': index,
      'plugin_name': pluginName,
      'cause_type': cause.runtimeType.toString(),
      'cause_message': cause.toString(),
      'message': toString(),
    };
  }

  @override
  String toString() {
    return 'resolver plugin failed at index $index ($pluginName): $cause';
  }
}

class Resolver {
  final List<ResolvePlugin> _plugins = <ResolvePlugin>[];

  void addPlugin(ResolvePlugin plugin) {
    _plugins.add(plugin);
  }

  Future<ResolveContext> resolve(Transaction tx) async {
    final context = ResolveContext(transaction: tx);
    for (final input in tx.data.inputs) {
      if (input['\$kind'] == 'UnresolvedObject') {
        context.unresolvedInputs.add(input);
      }
    }

    for (var i = 0; i < _plugins.length; i++) {
      final plugin = _plugins[i];
      final pluginName = '$plugin';
      try {
        await plugin(context);
      } catch (e) {
        throw ResolverPluginError(index: i, pluginName: pluginName, cause: e);
      }
    }

    return context;
  }
}

class CachingExecutor {
  CachingExecutor(this.client);

  final dynamic client;
  final Map<String, Map<String, dynamic>> _cache = <String, Map<String, dynamic>>{};

  Future<Map<String, dynamic>> executeTransaction(Transaction tx) async {
    final key = tx.buildBase64();
    final cached = _cache[key];
    if (cached != null) {
      return cached;
    }

    Map<String, dynamic> result;
    try {
      final out = await client.call('sui_executeTransactionBlock', [key]);
      result = Map<String, dynamic>.from(out as Map);
    } catch (_) {
      final out = await client.execute('sui_executeTransactionBlock', [key]);
      result = Map<String, dynamic>.from(out as Map);
    }
    _cache[key] = result;
    return result;
  }
}

class SerialExecutor {
  SerialExecutor(this.executor);

  final CachingExecutor executor;

  Future<List<Map<String, dynamic>>> execute(List<Transaction> txs) async {
    final out = <Map<String, dynamic>>[];
    for (final tx in txs) {
      out.add(await executor.executeTransaction(tx));
    }
    return out;
  }
}

class ParallelExecutor {
  ParallelExecutor(this.executor, {this.maxWorkers = 4});

  final CachingExecutor executor;
  final int maxWorkers;

  Future<List<Map<String, dynamic>>> execute(List<Transaction> txs) async {
    final out = List<Map<String, dynamic>?>.filled(txs.length, null, growable: false);
    var next = 0;

    Future<void> worker() async {
      while (true) {
        final i = next;
        next += 1;
        if (i >= txs.length) {
          return;
        }
        out[i] = await executor.executeTransaction(txs[i]);
      }
    }

    await Future.wait(List.generate(maxWorkers, (_) => worker()));
    return out.whereType<Map<String, dynamic>>().toList();
  }
}

File: py-sdks/src/pysdks/sui/transactions.py
"""Sui transaction builder and execution baseline."""

from __future__ import annotations

import base64
import json
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass, field
from threading import Lock
from typing import Any, Callable, Dict, List, Optional, Protocol

_SYSTEM_STATE_OBJECT_ID = "0x5"
_STAKE_REQUEST_TARGET = "0x3::sui_system::request_add_stake"
_UNSTAKE_REQUEST_TARGET = "0x3::sui_system::request_withdraw_stake"


@dataclass
class ObjectRef:
    object_id: str
    digest: str
    version: int


class Inputs:
    @staticmethod
    def pure(value: bytes) -> Dict[str, Any]:
        return {"$kind": "Pure", "Pure": {"bytes": base64.b64encode(value).decode()}}

    @staticmethod
    def object_ref(ref: ObjectRef) -> Dict[str, Any]:
        return {
            "$kind": "Object",
            "Object": {
                "$kind": "ImmOrOwnedObject",
                "ImmOrOwnedObject": {
                    "objectId": ref.object_id,
                    "digest": ref.digest,
                    "version": ref.version,
                },
            },
        }

    @staticmethod
    def shared_object_ref(object_id: str, mutable: bool, initial_shared_version: int) -> Dict[str, Any]:
        return {
            "$kind": "Object",
            "Object": {
                "$kind": "SharedObject",
                "SharedObject": {
                    "objectId": object_id,
                    "mutable": mutable,
                    "initialSharedVersion": initial_shared_version,
                },
            },
        }


class TransactionCommands:
    @staticmethod
    def move_call(target: str, args: List[Dict[str, Any]], type_args: Optional[List[str]] = None) -> Dict[str, Any]:
        pkg, mod, fn = (target.split("::") + ["", "", ""])[:3]
        return {
            "$kind": "MoveCall",
            "MoveCall": {
                "package": pkg,
                "module": mod,
                "function": fn,
                "arguments": args,
                "typeArguments": type_args or [],
            },
        }

    @staticmethod
    def split_coins(coin: Dict[str, Any], amounts: List[Dict[str, Any]]) -> Dict[str, Any]:
        return {"$kind": "SplitCoins", "SplitCoins": {"coin": coin, "amounts": amounts}}

    @staticmethod
    def transfer_objects(objects: List[Dict[str, Any]], address: Dict[str, Any]) -> Dict[str, Any]:
        return {"$kind": "TransferObjects", "TransferObjects": {"objects": objects, "address": address}}

    @staticmethod
    def merge_coins(destination: Dict[str, Any], sources: List[Dict[str, Any]]) -> Dict[str, Any]:
        return {"$kind": "MergeCoins", "MergeCoins": {"destination": destination, "sources": sources}}

    @staticmethod
    def publish(modules: List[bytes], dependencies: List[str]) -> Dict[str, Any]:
        encoded_modules = [base64.b64encode(m).decode() for m in modules]
        return {"$kind": "Publish", "Publish": {"modules": encoded_modules, "dependencies": dependencies}}

    @staticmethod
    def upgrade(modules: List[bytes], dependencies: List[str], package_id: str, ticket: Dict[str, Any]) -> Dict[str, Any]:
        encoded_modules = [base64.b64encode(m).decode() for m in modules]
        return {
            "$kind": "Upgrade",
            "Upgrade": {
                "modules": encoded_modules,
                "dependencies": dependencies,
                "package": package_id,
                "ticket": ticket,
            },
        }

    @staticmethod
    def make_move_vec(type_tag: Optional[str], elements: List[Dict[str, Any]]) -> Dict[str, Any]:
        return {"$kind": "MakeMoveVec", "MakeMoveVec": {"type": type_tag, "elements": elements}}


@dataclass
class TransactionData:
    sender: str = ""
    expiration: Any = None
    gas_data: Dict[str, Any] = field(default_factory=dict)
    inputs: List[Dict[str, Any]] = field(default_factory=list)
    commands: List[Dict[str, Any]] = field(default_factory=list)


class Transaction:
    def __init__(self, client: Any = None):
        self.data = TransactionData(gas_data={})
        self.client = client

    @property
    def commands(self):
        return self.data.commands

    @property
    def inputs(self):
        return self.data.inputs

    def set_sender(self, sender: str) -> None:
        self.data.sender = sender

    def set_sender_if_not_set(self, sender: str) -> None:
        if not self.data.sender:
            self.set_sender(sender)

    def set_gas_budget(self, budget: int) -> None:
        self.data.gas_data["budget"] = str(budget)

    def set_gas_budget_if_not_set(self, budget: int) -> None:
        if not self.data.gas_data.get("budget"):
            self.set_gas_budget(budget)

    def set_expiration(self, expiration: Any) -> None:
        self.data.expiration = expiration

    def set_gas_price(self, price: int) -> None:
        self.data.gas_data["price"] = str(price)

    def set_gas_owner(self, owner: str) -> None:
        self.data.gas_data["owner"] = owner

    def set_gas_payment(self, payments: List[Dict[str, Any]]) -> None:
        self.data.gas_data["payment"] = payments

    def gas(self) -> Dict[str, Any]:
        return {"$kind": "GasCoin", "GasCoin": True}

    def add_input(self, arg: Dict[str, Any]) -> Dict[str, Any]:
        self.data.inputs.append(arg)
        return {"$kind": "Input", "Input": len(self.data.inputs) - 1}

    def object(self, value: str | Dict[str, Any]) -> Dict[str, Any]:
        if isinstance(value, str):
            return self.add_input({"$kind": "UnresolvedObject", "UnresolvedObject": {"objectId": value}})
        if value.get("$kind") == "Input":
            return value
        return self.add_input(value)

    def pure(self, value: bytes) -> Dict[str, Any]:
        return self.add_input(Inputs.pure(value))

    def add_command(self, cmd: Dict[str, Any]) -> Dict[str, Any]:
        self.data.commands.append(cmd)
        return {"$kind": "Result", "Result": len(self.data.commands) - 1}

    def move_call(self, target: str, args: List[Dict[str, Any]], type_args: Optional[List[str]] = None) -> Dict[str, Any]:
        return self.add_command(TransactionCommands.move_call(target, args, type_args))

    def transfer_objects(self, objects: List[Dict[str, Any]], address: Dict[str, Any]) -> Dict[str, Any]:
        return self.add_command(TransactionCommands.transfer_objects(objects, address))

    def split_coins(self, coin: Dict[str, Any], amounts: List[Dict[str, Any]]) -> Dict[str, Any]:
        return self.add_command(TransactionCommands.split_coins(coin, amounts))

    def merge_coins(self, destination: Dict[str, Any], sources: List[Dict[str, Any]]) -> Dict[str, Any]:
        return self.add_command(TransactionCommands.merge_coins(destination, sources))

    def publish(self, modules: List[bytes], dependencies: List[str]) -> Dict[str, Any]:
        return self.add_command(TransactionCommands.publish(modules, dependencies))

    def upgrade(
        self,
        modules: List[bytes],
        dependencies: List[str],
        package_id: str,
        ticket: Dict[str, Any],
    ) -> Dict[str, Any]:
        return self.add_command(TransactionCommands.upgrade(modules, dependencies, package_id, ticket))

    def publish_upgrade(
        self,
        modules: List[bytes],
        dependencies: List[str],
        package_id: str,
        ticket: Dict[str, Any],
    ) -> Dict[str, Any]:
        return self.upgrade(modules=modules, dependencies=dependencies, package_id=package_id, ticket=ticket)

    def custom_upgrade(
        self,
        modules: List[bytes],
        dependencies: List[str],
        package_id: str,
        ticket: Dict[str, Any],
    ) -> Dict[str, Any]:
        return self.upgrade(modules=modules, dependencies=dependencies, package_id=package_id, ticket=ticket)

    def make_move_vec(self, type_tag: Optional[str], elements: List[Dict[str, Any]]) -> Dict[str, Any]:
        return self.add_command(TransactionCommands.make_move_vec(type_tag, elements))

    def transfer_sui(self, recipient: str, amount: int) -> Dict[str, Any]:
        split_result = self.split_coins(self.gas(), [self.pure(self._u64_bytes(amount))])
        return self.transfer_objects([split_result], self.pure(recipient.encode("utf-8")))

    def split_coin_equal(self, coin: Dict[str, Any], split_count: int, amount_per_split: int) -> Dict[str, Any]:
        if split_count <= 0:
            raise ValueError("split_count must be positive")
        amounts = [self.pure(self._u64_bytes(amount_per_split)) for _ in range(split_count)]
        return self.split_coins(coin, amounts)

    def split_coin_and_return(self, coin: Dict[str, Any], amount: int, recipient: str) -> Dict[str, Any]:
        split_result = self.split_coins(coin, [self.pure(self._u64_bytes(amount))])
        self.transfer_objects([split_result], self.pure(recipient.encode("utf-8")))
        return split_result

    def stake_coin(
        self,
        *,
        coins: List[str | Dict[str, Any]],
        validator_address: str,
        amount: Optional[int] = None,
        system_state_object_id: str = _SYSTEM_STATE_OBJECT_ID,
    ) -> Dict[str, Any]:
        coin_args = [self.object(coin) for coin in coins]
        coins_vec = self.make_move_vec(None, coin_args)
        amount_arg = self.pure(self._option_u64_bytes(amount))
        return self.move_call(
            _STAKE_REQUEST_TARGET,
            [
                self.object(system_state_object_id),
                coins_vec,
                amount_arg,
                self.pure(validator_address.encode("utf-8")),
            ],
        )

    def unstake_coin(
        self,
        *,
        staked_coin: str | Dict[str, Any],
        system_state_object_id: str = _SYSTEM_STATE_OBJECT_ID,
    ) -> Dict[str, Any]:
        return self.move_call(
            _UNSTAKE_REQUEST_TARGET,
            [self.object(system_state_object_id), self.object(staked_coin)],
        )

    def build(self) -> bytes:
        payload = {
            "Sender": self.data.sender,
            "Expiration": self.data.expiration,
            "GasData": self.data.gas_data,
            "Inputs": self.data.inputs,
            "Commands": self.data.commands,
        }
        return json.dumps(payload).encode()

    def build_base64(self) -> str:
        return base64.b64encode(self.build()).decode()

    def serialize(self) -> str:
        return self.build().decode()

    def get_transaction_data(self) -> Dict[str, Any]:
        return {
            "Sender": self.data.sender,
            "Expiration": self.data.expiration,
            "GasData": self.data.gas_data,
            "Inputs": self.data.inputs,
            "Commands": self.data.commands,
        }

    def deferred_execution(self) -> Dict[str, Any]:
        return {"sender": self.data.sender, "tx_bytes": self.build_base64()}

    def execute(
        self,
        client: Any = None,
        signatures: Optional[List[str]] = None,
        options: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        active_client = self._resolve_client(client)
        tx_bytes = self.build_base64()
        if signatures is None and options is None:
            return active_client.call("sui_executeTransactionBlock", [tx_bytes])
        return active_client.call("sui_executeTransactionBlock", [tx_bytes, signatures or [], options or {}])

    def inspect_all(self, client: Any = None, sender: Optional[str] = None) -> Dict[str, Any]:
        active_client = self._resolve_client(client)
        active_sender = sender or self.data.sender
        return active_client.call("sui_devInspectTransactionBlock", [active_sender, self.build_base64()])

    def inspect_for_cost(self, client: Any = None, sender: Optional[str] = None) -> Dict[str, int]:
        result = self.inspect_all(client=client, sender=sender)
        summary = ((result or {}).get("result") or {}).get("effects", {}).get("gasUsed", {})
        computation = int(summary.get("computationCost", 0))
        storage = int(summary.get("storageCost", 0))
        rebate = int(summary.get("storageRebate", 0))
        return {
            "computation_cost": computation,
            "storage_cost": storage,
            "storage_rebate": rebate,
            "total_cost": computation + storage - rebate,
        }

    @classmethod
    def from_serialized(cls, serialized: str) -> "Transaction":
        raw = base64.b64decode(serialized) if not serialized.startswith("{") else serialized.encode()
        payload = json.loads(raw.decode())
        tx = cls()
        tx.data.sender = payload.get("Sender", "")
        tx.data.expiration = payload.get("Expiration")
        tx.data.gas_data = payload.get("GasData", {})
        tx.data.inputs = payload.get("Inputs", [])
        tx.data.commands = payload.get("Commands", [])
        return tx

    def _resolve_client(self, client: Any = None) -> Any:
        active_client = client or self.client
        if active_client is None:
            raise ValueError("client is required for execution and inspection")
        return active_client

    @staticmethod
    def _u64_bytes(value: int) -> bytes:
        if value < 0:
            raise ValueError("u64 value must be non-negative")
        if value > (1 << 64) - 1:
            raise ValueError("u64 value out of range")
        return value.to_bytes(8, "little")

    @staticmethod
    def _option_u64_bytes(value: Optional[int]) -> bytes:
        if value is None:
            return b"\x00"
        return b"\x01" + Transaction._u64_bytes(value)


@dataclass
class ResolveContext:
    transaction: Transaction
    unresolved_inputs: List[Dict[str, Any]] = field(default_factory=list)


class ResolvePlugin(Protocol):
    def __call__(self, context: ResolveContext) -> None:
        ...


class Resolver:
    def __init__(self):
        self._plugins: List[ResolvePlugin] = []

    def add_plugin(self, plugin: ResolvePlugin) -> None:
        self._plugins.append(plugin)

    def resolve(self, tx: Transaction) -> ResolveContext:
        context = ResolveContext(transaction=tx)
        for inp in tx.data.inputs:
            if inp.get("$kind") == "UnresolvedObject":
                context.unresolved_inputs.append(inp)
        for plugin in self._plugins:
            plugin(context)
        return context


class CachingExecutor:
    def __init__(self, client: Any):
        self.client = client
        self._cache: Dict[str, Dict[str, Any]] = {}
        self._lock = Lock()

    def execute_transaction(self, tx: Transaction) -> Dict[str, Any]:
        key = tx.build_base64()
        with self._lock:
            if key in self._cache:
                return self._cache[key]
        result = self.client.call("sui_executeTransactionBlock", [key])
        with self._lock:
            self._cache[key] = result
        return result


class SerialExecutor:
    def __init__(self, executor: CachingExecutor):
        self.executor = executor

    def execute(self, txs: List[Transaction]) -> List[Dict[str, Any]]:
        out: List[Dict[str, Any]] = []
        for tx in txs:
            out.append(self.executor.execute_transaction(tx))
        return out


class ParallelExecutor:
    def __init__(self, executor: CachingExecutor, max_workers: int = 4):
        self.executor = executor
        self.max_workers = max_workers

    def execute(self, txs: List[Transaction]) -> List[Dict[str, Any]]:
        with ThreadPoolExecutor(max_workers=self.max_workers) as pool:
            futures = [pool.submit(self.executor.execute_transaction, tx) for tx in txs]
            return [f.result() for f in futures]
